<a href="https://colab.research.google.com/github/kmongsil1105/colab_ipynb/blob/main/AI%EC%9C%B5%ED%95%A9_%EA%B8%B0%EC%B4%88(CNN(%ED%95%A9%EC%84%B1%EA%B3%B1%EC%8B%A0%EA%B2%BD%EB%A7%9D).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 합성곱 신경망 (CNN)

### 텐서플로 임포트하기

In [ ]:
#!pip install tensorflow-gpu==2.0.0-rc1
import tensorflow as tf

from tensorflow.keras import datasets, layers, models

### MNIST 데이터셋 다운로드하고 준비하기

In [ ]:
(train_images, train_labels), (test_images, test_labels) = datasets.mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
test_images = test_images.reshape((10000, 28, 28, 1))

# 픽셀 값을 0~1 사이로 정규화합니다.
train_images, test_images = train_images / 255.0, test_images / 255.0

### 합성곱 층 만들기

In [ ]:
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1))) # 32 ==> feature map
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

In [ ]:
# 지금까지 모델의 구조를 출력
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_3 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 13, 13, 32)       0         
 2D)                                                             
                                                                 
 conv2d_4 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 conv2d_5 (Conv2D)           (None, 3, 3, 64)          36928     
                                                                 
Total params: 55,744
Trainable params: 55,744
Non-trai

### 마지막에 Dense 층 추가하기 ==> MLP

In [ ]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dropout(0.2)),  # "과적합" 발생으로 추가시킴==> 다시 처음부터 실행!!!
model.add(layers.Dense(10, activation='softmax'))

In [ ]:
# 최종 모델의 구조를 확인
model.summary() # MNIST는 0~9(10개)를 분류하므로.. 마지막 layer의 퍼셉트론이 10개!!!

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_3 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 13, 13, 32)       0         
 2D)                                                             
                                                                 
 conv2d_4 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 conv2d_5 (Conv2D)           (None, 3, 3, 64)          36928     
                                                                 
 flatten_1 (Flatten)         (None, 576)              

### 모델 컴파일과 훈련하기

In [ ]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 58s 30ms/step - loss: 0.1856 - accuracy: 0.9431
Epoch 2/5
1875/1875 [==============================] - 57s 30ms/step - loss: 0.0587 - accuracy: 0.9829
Epoch 3/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0411 - accuracy: 0.9876
Epoch 4/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0328 - accuracy: 0.9900
Epoch 5/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0250 - accuracy: 0.9926


### 모델 평가

In [ ]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

313/313 - 3s - loss: 0.0323 - accuracy: 0.9911 - 3s/epoch - 9ms/step


In [ ]:
print(test_acc)

0.991100013256073


학습 accuracy: 0.9937 / 테스트 accuracy: 0.9903 

==> 학습 정확도 > 테스트 정확도 :: "과적합" 발생

==> DropOut 적용

학습 accuracy : 0.9926 / 테스트 accuracy: 0.9911    loss: 0.0323 

==> 학습 정확도와 테스트 정확도의 차이가 줄어듬!!!
